### PR TITLE
AddDependency will find specified types not already found

### PR DESCRIPTION
**Pull request by pway99**
AddDependency Recipe either responds to a marker put into the ExecutionContext by other visitors or finds specified types which have not yet been found.  Fixes https://github.com/openrewrite/rewrite-testing-frameworks/issues/116

This change 
 - adds a new HasTypes recipe which will add a RecipeSearchResultMarker to any J.CompilationUnit having any one of the supplied types
 - removes `onlyIfUsing` from the `AddDependencyVisitor`

### CODE REVIEW

File: rewrite-core/src/main/java/org/openrewrite/Recipe.java [hunk 1]
```diff
@@ -149,6 +149,13 @@ public final RecipeDescriptor getDescriptor() {
     @JsonIgnore
     private final List<Recipe> recipeList = new ArrayList<>();
 
+    /**
+     * Invoked once after recipe has finally completed
+     * @param executionContext {@link ExecutionContext} from recipe execution
+     */
+    @Incubating(since = "7.2.0-SNAPSHOT")
```
Comment (sambsnyd) [inline rewrite-core/src/main/java/org/openrewrite/Recipe.java:156]:
```suggestion
    @Incubating(since = "7.2.0")
```

**Comment by jkschneider** [comment]
I think `afterRecipe` is logically equivalent to `getApplicableTest`.

**Comment by pway99** [comment]
> I think `afterRecipe` is logically equivalent to `getApplicableTest`.

Thanks for the suggestion, @jkschneider.  I have re-factored the code to use`getApplicableTest` for this scenario by adding the `ExecutionContext` as a parameter.  The result is simpler, and validation happens first rather than requiring another full pass over all source files.

**Comment by jkschneider** [comment]
Really well done @pway99 !

**Comment by pway99** [comment]
also Fixes: https://github.com/openrewrite/rewrite/issues/401

**Pull request merged**